### DISCUSSION TITLE
How do I ensure a specific string is always prepended to the BEGINNING of the final selector in my custom variant?

### DISCUSSION THREAD

**Discussion by AaronBeaudoin** [unocss/unocss#1772] [Q&A] [answered]
I'm trying to create a custom variant that will prepend `html[data-input=mouse]` to the final selector whenever the _existing_ built-in `hover:` or `group-hover:` variants are being used. The problem I'm running into is that no matter what I do, the final selector still has `.group:hover` _before_ `html[data-input=mouse]` which is **not** what I want.

If I specify classes in my code like `hover:bg-white group-hover:bg-white`, I want CSS like this:

```css
html[data-input=mouse] .hover\:bg-white:hover { ... }
html[data-input=mouse] .group:hover .group-hover\:bg-white { ... }
```

_Instead_, I'm getting CSS like this:

```css
html[data-input=mouse] .hover\:bg-white:hover { ... }
.group:hover html[data-input=mouse] .group-hover\:bg-white { ... }
```

Here is my custom preset which I am using to add the variant:

```js
{
  name: "custom-hover",
  variants: [
    (matcher, context) => {
      const original = context.rawSelector;
      const match = original.includes("hover:") || original.includes("group-hover:");
      if (!match) return matcher;

      return {
        selector: _ => `html[data-input=mouse] ${_}`,
        matcher: matcher
      };
    }
  ]
}
```

I am listing it _after_ the built-in default preset, so I would expect it to be processed last.

**Comment by chu121su12** [comment, marked as answer]
Variants now uses 3-part sections: prefix, selector and pseudo. Also there's onion skin style resolution in variant.

[See example
](https://unocss.antfu.me/play/?html=DwEwlgbgBAxgNgQwM5ILwCIBQUdQBYD2EApgE4BcARgOYC0A7nmAC7FTWkECuADrYSQo0GTVpnQA%2BTEA&config=JYWwDg9gTgLgBAbzgEwKYDNgDtUGEJaYDmcAvnOlBCHAOQCuWEAxgM6u0BQnqAHpLBQYAhvQA28NJhz5CwIgAoEnOHABuwqMGFYYrAFxwA2itWJTZuCGExmACwXXbd1FAA0cZgRh8YASnNLSy8sVnhoeWxhMTgAXk9vXwA6KGEAdwBlVDFUZhhoAG4LIJCwqxt7OLgIoiixJOxmMXo0VgUAIjsINVd9doCAHwHqrVqsaIasJpbUNvaiKnowAFounqg%2BvyKgs2B0OAUAQid7AKhUGHooLHLnVyLiy3PL68CdsxOXDdv7VzdHoJ2HTIHKGBTYMD0GAeHC8fxxAB8b3eZlK4ShVVhMHBWEh-gBO2eVxuyhROySFIgUP%2BZKCYHOmF4hgABnYYCAxEZkDZhMsIVDYiAqaxUABdOAAEgQVJgSXpGGAvDgQzotFIzJptLIBLMpE1O1I2yCpGKJtUov%2BpD8QA&options=N4IgzgLgTglgxhEAuaBXApgXyA)

**Reply by AaronBeaudoin** [reply]
This worked perfectly, thank you! I wish this information was documented somewhere.
